### ISSUE TITLE
make generate takes very long

### ISSUE THREAD

**Issue by timebertt** [gardener/gardener#3388]
**How to categorize this issue?**

/area dev-productivity
/kind enhancement
/priority normal

**What would you like to be added**:

Make `make generate` faster.

**Ideas**
I think, [controller-tools](https://github.com/kubernetes-sigs/controller-tools) might be a bit faster than [code-generator](https://github.com/kubernetes/code-generator).
We could try, switching to controller-tools for the generations it supports.

Another idea is to parallelize different code generations. 
On my machine, `make generate` consistently utilizes about 50% cpu.
So at least for local runs, the time can maybe be split by half by parallelization.

**Why is this needed**:

`make generate` takes forever.
This is a very bad developer experience and also costs countless hours of CI time.

**Comment by stoyanr** [comment]
> the time can maybe be split by half by parallelization.

Do you have ideas how we could achieve this? Unfortunately, `go generate` doesn't seem to support parallelization (see also https://github.com/golang/go/issues/20520). We could probably split the single `go generate` run into multiple runs for specific  files / packages that then could be executed in parallel, but it would be harder (though not impossible) to do it without manually maintaining a list of packages to be go-generated somewhere.

Another idea that comes to mind is to parallelize only the runs within `hack/update-codegen.sh`, as this takes the most time by far. This should be easier, I could give it a quick try to see if the benefit is worth it.

Few other things that come to mind:
* The level of parallelisation should be configurable somewhere and people should be able to specify it depending on e.g. how many processors they have or how much CPU power they want to reserve for other tasks.
* Output of parallel code generation tasks would be garbled and unusable, so we should generally switch it off except for errors.

**Comment by timebertt** [comment]
> Another idea that comes to mind is to parallelize only the runs within `hack/update-codegen.sh`, as this takes the most time by far. This should be easier, I could give it a quick try to see if the benefit is worth it.

This was also the first step, I was thinking about.
Possible ideas for parallelizing would be to either do it in bash, or even write a small go wrapper using the code generator's go code directly.

> I think, [controller-tools](https://github.com/kubernetes-sigs/controller-tools) might be a bit faster than [code-generator](https://github.com/kubernetes/code-generator).
> We could try, switching to controller-tools for the generations it supports.

You can also give this one a shot if you want. Don't know if code generation using controller-tools is significantly faster, though.
As far as I can see, it also doesn't support all types of code generation we do.

**Comment by stoyanr** [comment]
> Possible ideas for parallelizing would be to either do it in bash, or even write a small go wrapper using the code generator's go code directly.

Let me fist do a small PoC in bash to see how much we are likely to gain.

> I think, controller-tools might be a bit faster than code-generator.

I would save this for last, to me at least it sounds less promising than the parallel execution.

**Comment by timebertt** [comment]
> > I think, controller-tools might be a bit faster than code-generator.
> 
> I would save this for last, to me at least it sounds less promising than the parallel execution.

Yes, this was not meant to be either or, we can probably do both if it is of any benefit.

**Comment by stoyanr** [comment]
For posterity, besides the improvements implemented in #3489 I also experimented with using [controller-tools](https://github.com/kubernetes-sigs/controller-tools). Here is what I found.

There is a single generator, `controller-gen`, and the only generation it can do that is alternative to `code-generator` is for deepcopy functions, if it's invoked like this:

```
  go run "${PROJECT_ROOT}"/vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go \
    object:headerFile="${PROJECT_ROOT}"/hack/LICENSE_BOILERPLATE.txt,year="${YEAR}" \
    paths="${PROJECT_ROOT}/$1"
```

Initially, I was seeing a huge difference in the performance, `controller-gen` was much faster than `deepcopy-gen`, so I decided to implement a change to `hack/update-codegen.sh` that uses it. This was before we introroduced `GO111MODULE=off`. With `GO111MODULE=off`, this difference dropped to a few negligible seconds, obviously the biggest part of it was because `controller-gen` was ignoring `GO111MODULE`.

Also, I found out a few issues with the deepcopy generation of `controller-gen`:
* It fails on types that embed `runtime.Object`, so for packages that contain such types (2 out of 9 in our case) I had to fallback to `deepcopy-gen`, since ignoring such types leads to other errors down the road.
* It fails with interface types, I had to explicitly ignore them by adding `// +kubebuilder:object:generate=false` on such types - not a big deal but an unnecessary complication.
* It generates slightly different deepcopy code - all still working, but `goimports` considers it to be improperly formatted due to a small glitch in the imports generation.

I managed to make it work only to realize on the next day that it was a waste of time after we discovered that `GO111MODULE=off` was mostly what we needed. I decided not to open a PR for it, it would only increase the complexity but bring no real performance improvement and so no value whatsoever.
